tcases-lib/src/main/java/org/cornutum/tcases/generator/io/GeneratorSetDocWriter.java

```diff
@@ -100,107 +98,89 @@ protected void writeGenerator( String function, ITestCaseGenerator generator)
+
+      MapBuilder
+        .of( FUNCTION_ATR, function)
+        .putIf( SEED_ATR, Optional.ofNullable( seed != null? String.valueOf( seed) : null))
+        .put( TUPLES_ATR, String.valueOf( generator.getDefaultTupleSize()))
+        .build(),
+      
+      () ->
+        {
+        for( TupleCombiner combiner : generator.getCombiners())
+          {
+          writeCombiner( combiner);
+          }
+        });
     }
 
   /**
    * Writes the given combiner definition.
    */
   protected void writeCombiner( TupleCombiner combiner)
     {
-    writer_.writeTagStart( COMBINE_TAG);
-    writer_.writeAttribute( TUPLES_ATR, String.valueOf( combiner.getTupleSize()));      
-    writer_.writeTagEnd();
-
-    writer_.indent();
-    
-    String[] included = combiner.getIncluded();
-    Arrays.sort( included);
-    for( int i = 0; i < included.length; i++)
-      {
-      writeIncluded( included[i]);
-      }
-    
-    String[] excluded = combiner.getExcluded();
-    Arrays.sort( excluded);
-    for( int i = 0; i < excluded.length; i++)
-      {
-      writeExcluded( excluded[i]);
-      }
-
-    for( Iterator<TupleRef> onceTuples = combiner.getOnceTuples(); onceTuples.hasNext(); )
-      {
-      writeOnceTuple( onceTuples.next());
-      }
-    
-    writer_.unindent();
-    
-    writer_.writeElementEnd( COMBINE_TAG);
+    writer_.writeElement(
+      COMBINE_TAG,
+      MapBuilder.of( TUPLES_ATR, String.valueOf( combiner.getTupleSize())).build(),
+      () ->
+        {
+        Arrays.stream( combiner.getIncluded())
+          .sorted()
+          .forEach( included -> writeIncluded( included));
+        
+        Arrays.stream( combiner.getExcluded())
+          .sorted()
+          .forEach( excluded -> writeExcluded( excluded));
+
+        toStream( combiner.getOnceTuples()).forEach( tuple -> writeOnceTuple( tuple));
+        });
     }
 
   /**
    * Writes the given once-only tuple definition.
    */
   protected void writeOnceTuple( TupleRef tuple)
     {
-    writer_.writeElementStart( ONCE_TAG);
-    writer_.indent();
-    for( Iterator<VarBinding> varBindings = tuple.getVarBindings(); varBindings.hasNext(); )
-      {
-      writeVarBinding( varBindings.next());
-      }
-    writer_.unindent();
-    writer_.writeElementEnd( ONCE_TAG);
+    writer_.writeElement(
+      ONCE_TAG,
+      () ->
+        {
+        toStream( tuple.getVarBindings()).forEach( varBinding -> writeVarBinding( varBinding));
+        });
     }
 
   /**
    * Writes the given variable binding definition.
    */
   protected void writeVarBinding( VarBinding binding)
     {
-    writer_.writeTagStart( VAR_TAG);
-    writer_.writeAttribute( NAME_ATR, binding.getVar());      
-    writer_.writeAttribute( VALUE_ATR, String.valueOf( binding.getValue()));      
-    writer_.writeEmptyElementEnd();
+    writer_.writeElement(
+      VAR_TAG,
+      MapBuilder.of( NAME_ATR, binding.getVar()).put( VALUE_ATR, String.valueOf( binding.getValue())).build());
```

Comment:
personally I find the combination of of() and put() weird. new MapBuilder().put(...).put(...) does not leave readers to wonder.